### DISCUSSION TITLE
Error handling problem when disconnecting a device

### DISCUSSION THREAD

**Discussion by vascolp** [pybricks/support#558] [Q&A]
Hello,

I think I might have found something wrong in version v3.1.0c1 and also in the just released v3.1.0.
At least, in v3.1.0b1 things make much more sense to me...
It is related with disconnecting a device from the TechnicHub and exception handling.
Consider the below program to run in a TechnicHub with a Motor in Port.A.
It asks for a remote (white blinking light in th hub), and then, if you press a remote button it prints the device object.
With this program, if you disconnect the device, the program should print an error and go back to the main cycle.
But with versions v3.1.0c1 and v3.1.0 it crashes.

Ah! Just found out that it only happens with Motor, it does not happen with DCMotor!

Here is the output for v3.1.0b1, assuming you have a motor connected to Port.A when it starts.
You have to disconnect the Motor to get an exception and then you can connect it back with no problems.
```
('technichub', '3.1.0b1', 'v3.1.0b1 on 2021-09-21')
Button pressed: Motor(Port.A, Direction.CLOCKWISE)
Button pressed: Motor(Port.A, Direction.CLOCKWISE)
Exception!
Button pressed: Motor(Port.A, Direction.CLOCKWISE)
```

Here is the output for v3.1.0c1 (similar to v3.1.0), assuming you have a motor connected to Port.A when it starts.
In this case, as soon as you disconnect the motor the program crashes.
I also noticed that if you start the program without the motor connected, the Exception is properly handled, but if you connect the motor, press a remote button and then disconnect the motor, it crashes.
Notice: the error looks like the error associated with the hub button but it happens when you desconnect the Motor cable. I never press the hub button.
```
('technichub', '3.1.0c1', 'v3.1.0c1 on 2021-11-19')
Button pressed: Motor(Port.A, Direction.CLOCKWISE)
Button pressed: Motor(Port.A, Direction.CLOCKWISE)
Traceback (most recent call last):
  File "tt.py", line 16, in <module>
SystemExit: stop button pressed
```

```python
# Test program
from pybricks.hubs import TechnicHub
from pybricks.pupdevices import Remote, Motor
from pybricks.parameters import Port, Direction, Stop, Button, Color
from pybricks.tools import wait
hub=TechnicHub()
hub.light.blink(Color.WHITE, (200,200, 100, 400))
rem = Remote()
hub.light.on(Color.WHITE)

from pybricks import version
print(version)

while True:
    try:
        wait(100)
        pressed = rem.buttons.pressed()
        if len(pressed)!=1:
            continue

        d=Motor(Port.A)
        print('Button pressed:', d)
    except Exception as e:
        print('Exception!')
        wait(1000)
        hub.light.blink(Color.ORANGE, (100,50))
        wait(1000)
        hub.light.on(Color.WHITE)
```

Thanks and best regards!
VascoLP

**Comment by BertLindeman** [comment]
Hi vascolp,

Maybe you now need to change the use of center button?
Pressing the center button is acting like the stop button.
[edit] (Or I might have mis-interpreted your report about: which button)

In the beta examples I see this:
``` python
# Using the stop button during your program
from pybricks.hubs import TechnicHub
from pybricks.parameters import Color, Button
from pybricks.tools import wait, StopWatch

# Initialize the hub.
hub = TechnicHub()

# Disable the stop button.
hub.system.set_stop_button(None)

# Check the button for 5 seconds.
watch = StopWatch()
while watch.time() < 5000:

    # Set light to green if pressed, else red.
    if hub.button.pressed():
        hub.light.on(Color.GREEN)
    else:
        hub.light.on(Color.RED)

# Enable the stop button again.
hub.system.set_stop_button(Button.CENTER)

# Now you can press the stop button as usual.
wait(5000)
```

**Comment by vascolp** [comment]
Thank you for your answer,  I am aware of the set_stop_button method. In fact that was the last thing I removed from my test program.
I might have not explained well, but the problem has nothing to do with the hub button or any other button. The problem arises when you disconnect the motor cable from the hub. 
Yes, the error looks like the error associated to the hub button. Which is even stranger, because it happens when disconnecting a cable...

Ah! Just found out that it only happens with Motor, it does not happen with DCMotor!

**Comment by laurensvalk** [comment]
This is the expected behavior for this release.

When you unplug a motor while it is running, the System Exit exception is raised. This also happens if you press the stop button, and there’s only one generic ‘stop button was pressed’ message right now.

We did this to make sure the hub would cleanly turn off all the motors in case of unhandled problems, like unplugging a cable.

A nicer solution that works without the global System Exit has already been developed. It will raise an OSError with errno=ENODEV for that particular motor only. This will be part of a future release. Probably the next beta.

**Reply by laurensvalk** [reply]
Fun fact: I thought we might just get away with this hack without anyone noticing, but it’s great to see all releases being this carefully scrutinized. Thanks @vascolp!

**Comment by vascolp** [comment]
Hi Laurensvalk,

Ok, I can live with that...  but not convinced :-)
It only happens with Motor, not DCMotor, and it is not catchable? 
Definitely not the best solution... you will fix it in a future release, so maybe we both agree on that :-)

Is the System Exit never catchable?
Does this hack happens inside a wait() call? It looks like that...
Thanks!
PS: The reason I want to be able to catch is this: [RemoteBlaBla](https://github.com/vascolp/RemoteBlaBla) . It is a solution for technic hub without the need for a mobile phone.  And there, connecting and disconnecting devices is part of the configuration process...

**Reply by laurensvalk** [reply]
If you want to try the latest version already and give feedback, [here's how you can do that](https://pybricks.com/install/technic-boost-city/#installing-the-latest-build-advanced).

> Is the System Exit never catchable?

It should be possible to catch it. Let us know if this is not working in the latest version :)

And for detecting devices, see also the [PUPDevices class](https://docs.pybricks.com/en/latest/iodevices/pupdevice.html), which may be helpful.

**Comment by vascolp** [comment]
Thank you for your answer.
About releases, I don´t really have the time to do further tests right now, but I tested with all latest releases available [here](https://github.com/pybricks/pybricks-micropython/releases).
So, with v3.1.0, not being able to catch System Exit  seems to be a fact (at least the one generated when we disconnect a Motor device, as we talked).

I am already using PUPDevice class, it works fine with Motor devices. I noted some issues with DCMotor devices (ID 1 I think, the one from batman LEGO set 76112), specially when battery charge is low.  Sometimes the code would block until we disconnect the device. I think there was a detection problem related with a voltage level (I might have read about this somewere around here), but I had this situation much more often in v3.1.0b1, it seems v3.1.0c1 fixed it.

**Comment by vascolp** [comment]
Hi Pybrickers!

Sorry  to get back to this problem, but I think there is really something wrong when you disconnect a device.
If you run the code below with a motor attached to port C, it should be able to catch the generated 
SystemExit exception and go back to the loop.
What happens after disconnecting the motor is that the wait will always generate a SystemExit exception.
It seems that after disconnecting the motor the wait becomes confused...
```
from pybricks.hubs import ThisHub
from pybricks.pupdevices import Motor
from pybricks.parameters import Port
from pybricks.tools import wait

hub=ThisHub()
hub.system.set_stop_button(None)

d=Motor(Port.C)

try:
    while True:
        try:
            try:
                wait(1500)
            except SystemExit as e:
                print('SystemExit2!')
            print('Hello!')
        except SystemExit as e:
            print('SystemExit!')
except Exception as e:
    print('AAAA')
```

**Reply by dlech** [reply]
Does this still happen with the latest firmware builds from CI? https://github.com/pybricks/pybricks-micropython/actions/workflows/build.yml?query=is%3Asuccess+branch%3Amaster

**Comment by vascolp** [comment]
Tried in this:  v3.1.0 on 2021-12-16

**Comment by vascolp** [comment]
I don´t know how to get a firmware version from there!

**Reply by dlech** [reply]
See https://pybricks.com/install/technic-boost-city/#installing-the-latest-build-advanced

**Comment by vascolp** [comment]
Cool!
Works fine on this one: v3.1.0-177-ga56fbbae on 2022-02-22

Thanks!
